src/assembly/manifest.ts

```diff
@@ -0,0 +1,166 @@
+import * as path from 'path';
+import {
+    AssemblyManifest,
+    Manifest,
+    ArtifactType,
+    AwsCloudFormationStackProperties,
+    ArtifactMetadataEntryType,
+} from '@aws-cdk/cloud-assembly-schema';
+import * as fs from 'fs-extra';
+import { CloudFormationTemplate } from '../cfn';
+import { ArtifactManifest, AssetManifestProperties, LogicalIdMetadataEntry } from 'aws-cdk-lib/cloud-assembly-schema';
+import { CloudAssembly, CloudFormationStackArtifact } from 'aws-cdk-lib/cx-api';
+import { AssetManifest, DockerImageManifestEntry, FileManifestEntry } from 'cdk-assets';
+import { StackManifest } from './stack';
+import { ConstructTree, StackAsset, StackMetadata } from './types';
+
+/**
+ * Reads a Cloud Assembly manifest
+ */
+export class AssemblyManifestReader {
+    public static readonly DEFAULT_FILENAME = 'manifest.json';
+
+    /**
+     * Reads a Cloud Assembly manifest from a file
+     */
+    public static fromFile(fileName: string): AssemblyManifestReader {
+        try {
+            const obj = Manifest.loadAssemblyManifest(fileName);
+            return new AssemblyManifestReader(path.dirname(fileName), obj);
+        } catch (e: any) {
+            throw new Error(`Cannot read manifest '${fileName}': ${e.message}`);
+        }
+    }
+
+    /**
+     * Reads a Cloud Assembly manifest from a file or a directory
+     * If the given filePath is a directory then it will look for
+     * a file within the directory with the DEFAULT_FILENAME
+     */
+    public static fromPath(filePath: string): AssemblyManifestReader {
+        let st;
+        try {
+            st = fs.statSync(filePath);
+        } catch (e: any) {
+            throw new Error(`Cannot read manifest at '${filePath}': ${e.message}`);
+        }
+        if (st.isDirectory()) {
+            return AssemblyManifestReader.fromFile(path.join(filePath, AssemblyManifestReader.DEFAULT_FILENAME));
+        }
+        return AssemblyManifestReader.fromFile(filePath);
+    }
+
+    /**
+     * The directory where the manifest was found
+     */
+    public readonly directory: string;
+
+    private readonly assembly: CloudAssembly;
+    private readonly stacks = new Map<string, CloudFormationStackArtifact>();
+    private readonly _stackManifests = new Map<string, StackManifest>();
+    private readonly tree: ConstructTree;
+
+    constructor(directory: string, private readonly manifest: AssemblyManifest) {
+        this.directory = directory;
+        this.assembly = new CloudAssembly(directory, {
+            // we don't need version checking / version checking would mean we would have to
+            // publish a new version of the library everytime the cdk version increases
+            skipVersionCheck: true,
```

Comment:
It's a json schema version check thing. So if we were using CDK v1.0.0 and someone were to upgrade to CDK v1.0.1 this would fail because we can only read <=v1.0.0. It's a generic version check. It's possible that they could do a major version upgrade which breaks backwards compatibility, but I think we have to handle major version changes regardless.

Reply:
Actually I realized this is not needed and was left over from something else. Removed!